### DISCUSSION TITLE
Small request performance

### DISCUSSION THREAD

**Discussion by avikivity** [minio/minio#13488] [Q&A] [answered]
As a test, I started minio on my workstation with a data directory off my /tmp (running tmpfs). I uploaded a 1GB object, set the bucket for public access, and stressed it with `wrk` limiting it to a 512 byte range:

```console
$ ./wrk -t 10 -c 300 -H 'Range: bytes=0-511'  http://localhost:9000/sdfff/obj
Running 10s test @ http://localhost:9000/sdfff/obj
  10 threads and 300 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    12.83ms   17.75ms 241.20ms   82.22%
    Req/Sec     7.85k     1.93k   63.37k    98.20%
  781768 requests in 10.10s, 772.39MB read
Requests/sec:  77404.96
Transfer/sec:     76.48MB
```

I got about 77k reqs/sec. minio was consuming 2600% cpu, and dividing the two numbers I get about 3k/sec/vcpu.

I performed another depth with iodepth 1 (again reading just 512 bytes):

```console
$ ./wrk  -t 1 -c 1 -H 'Range: bytes=0-511'  http://localhost:9000/sdfff/obj
Running 10s test @ http://localhost:9000/sdfff/obj
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   327.76us   86.13us   2.03ms   94.48%
    Req/Sec     3.09k   108.58     3.37k    69.00%
  30754 requests in 10.00s, 30.39MB read
Requests/sec:   3075.35
```

and observed the same 3k reqs/sec.

My question: does minio really require a minimum of 0.3ms to process a request? There were no network hops involved (single server), no I/O (tmpfs). Is there anything I can do to get more small-request performance?

**Comment by harshavardhana** [comment]
@avikivity are you using FS mode or erasure-coded mode here?

**Reply by avikivity** [reply]
I'm using the default, I assume it's FS-mode (I'm new to minio). This is a single-node cluster and I provided only one directory to the server.

**Reply by avikivity** [reply]
The objects appear in the directory without change, so I think this proves erasure coding is not used.

**Comment by harshavardhana** [comment]
> ```
>     Latency   327.76us   86.13us   2.03ms   94.48%
> ```

It shows that your latest average `327` microseconds, not `milli seconds`  @avikivity - are you reading something else? 

MinIO reads directly from the disk in O_DIRECT fashion, so it is going to give you the best that is possible for your disk.

**Reply by avikivity** [reply]
Yes it's 327 microseconds. The disk is tmpfs, so there is no disk latency involved, it's pure CPU work.

**Reply by avikivity** [reply]
And I'm concerned about the throughput per vcpu, not the latency here.

**Reply by harshavardhana** [reply]
> Yes it's 327 microseconds. The disk is tmpfs, so there is no disk latency involved, it's pure CPU work.

327 microseconds is perfectly okay @avikivity that is simply an overhead of net/http I don't see anything can be done there.

**Reply by avikivity** [reply]
That's very disappointing, getting 3k ops/sec from a vcpu for a trivial op is far below what I expected.

**Reply by avikivity** [reply]
I see people reached 11.5k ops/sec/vcpu with go: https://stackoverflow.com/questions/37714210/go-http-server-slow-benchmark-performance

**Reply by harshavardhana** [reply]
> That's very disappointing, getting 3k ops/sec from a vcpu for a trivial op is far below what I expected.

what would you expect?

**Reply by harshavardhana** [reply]
> I see people reached 11.5k ops/sec/vcpu with go: https://stackoverflow.com/questions/37714210/go-http-server-slow-benchmark-performance

That benchmark is just pure memory with static content - that is not a real-world number although good enough. 

We have to `open(),  readv()`  this adds latency - if you `strace` your process you will see the latencies coming from. `tmpfs` is no magic that still involves latency at various layers. 

AT this point I am not sure what you are really optimizing here 3k requests per CPU is completely okay, as a client you are always bottlenecked way before you hit this number.

**Reply by avikivity** [reply]
For such simple ops, with no extra hops to another server, around 20k ops/sec/vcpu, maybe even more. Of course if we have to hop to another server and perform checksums, and read from a real disk, performance would drop.

**Reply by avikivity** [reply]
You could cache open files, and use linux-aio or io-uring to batch system calls. In any case a system call is around 1-2 usec, and here there are 327 usecs.

I am trying to move my application from local NVMe to object store. My application can drive much more than 3k ops/sec/vcpu, so if minio is slower than that I'd need several minio vcpus per application vcpu.

**Reply by avikivity** [reply]
btw, `top` reported that most of the CPU was user time, not system time, so it's not system calls to blame here.

**Reply by harshavardhana** [reply]
> You could cache open files, and use linux-aio or io-uring to batch system calls. In any case a system call is around 1-2 usec, and here there are 327 usecs.

We do not cache Open files @avikivity we always do O_DIRECT - that is part of the design we will never rely on Linux page-cache. 

Can you share the information where you saw syscall is around `1-2usec` I don't see this information shared anywhere?

**Reply by harshavardhana** [reply]
> btw, `top` reported that most of the CPU was user time, not system time, so it's not system calls to blame here.

Again you didn't share the information - I can't see what you are seeing @avikivity - please provide some information - so that I can validate the claims on my end.

**Reply by avikivity** [reply]
> > You could cache open files, and use linux-aio or io-uring to batch system calls. In any case a system call is around 1-2 usec, and here there are 327 usecs.
> 
> We do not cache Open files @avikivity we always do O_DIRECT - that is part of the design we will never rely on Linux page-cache.

I meant caching open file descriptors, not the file data, so you don't need to call open again and again. I totally appreciate not using the page-cache - my application does the same with local storage.

> 
> Can you share the information where you saw syscall is around `1-2usec` I don't see this information shared anywhere?


Here's one [example](https://stackoverflow.com/questions/23599074/system-calls-overhead). I expect today syscalls are somewhat slower due to spectre/meltdown mitigations.

**Reply by avikivity** [reply]
> > btw, `top` reported that most of the CPU was user time, not system time, so it's not system calls to blame here.
> 
> Again you didn't share the information - I can't see what you are seeing @avikivity - please provide some information - so that I can validate the claims on my end.

Here's a perf run while the benchmark is running:

```console
$ perf stat -p `pgrep minio` -e cycles:u -e cycles:k
^C
 Performance counter stats for process id '1790137':

   226,964,171,537      cycles:u                                                    
    31,801,995,991      cycles:k                                                    

       2.390539216 seconds time elapsed
```

31/226 = 13% time in kernel code.

**Comment by harshavardhana** [comment, marked as answer]
This is interesting I will look more deeper and get back - there may be a top level overhead somewhere.

**Reply by avikivity** [reply]
Many thanks. I'm happy to provide more information as needed.

**Comment by harshavardhana** [comment]
Found some performance gains

Latest release
```
./wrk  -t 1 -c 1 http://192.168.203.5:9001/testbucket/wrk
Running 10s test @ http://192.168.203.5:9001/testbucket/wrk
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   304.35us  340.35us   5.05ms   96.12%
    Req/Sec     3.87k   209.15     4.30k    65.00%
  38532 requests in 10.00s, 36.16MB read
Requests/sec:   3852.56
Transfer/sec:      3.62MB
```

With some patches
```
./wrk  -t 1 -c 1 http://192.168.203.5:9001/testbucket/wrk
Running 10s test @ http://192.168.203.5:9001/testbucket/wrk
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   249.87us  211.88us   4.60ms   98.12%
    Req/Sec     4.33k   237.90     4.95k    71.29%
  43448 requests in 10.10s, 40.77MB read
Requests/sec:   4301.86
Transfer/sec:      4.04MB
```

Pure Go implementation of reading a file and returning seems to be giving 

```go
package main

import (
        "io"
        "net/http"
        "os"

        disk "github.com/minio/minio/internal/disk"
        xioutil "github.com/minio/minio/internal/ioutil"
)

func hello(w http.ResponseWriter, r *http.Request) {
        off := int64(0)
        length := int64(512)
        fr, err := os.Open("/run/fs/wrk")
        if err != nil {
                return
        }

        // Stat to get the size of the file at path.
        st, err := fr.Stat()
        if err != nil {
                fr.Close()
                return
        }

        // Verify if its not a regular file, since subsequent Seek is undefined.
        if !st.Mode().IsRegular() {
                fr.Close()
                return
        }

        // Seek to the requested offset.
        if off > st.Size() || off+length > st.Size() {
                fr.Close()
                return
        }

        alignment := off%xioutil.DirectioAlignSize == 0

        sr := io.NewSectionReader(fr, off, length)
        rd := struct {
                io.Reader
                io.Closer
        }{Reader: sr, Closer: closeWrapper(func() error {
                if !alignment || off+length%xioutil.DirectioAlignSize != 0 {
                        // invalidate page-cache for unaligned reads.
                        disk.FadviseDontNeed(fr)
                }
                return fr.Close()
        })}
        defer rd.Close()
        io.Copy(w, rd)
}

// closeWrapper converts a function to an io.Closer
type closeWrapper func() error

// Close calls the wrapped function.
func (c closeWrapper) Close() error {
        return c()
}

func main() {
        http.HandleFunc("/", hello)
        http.ListenAndServe(":8000", nil)
}
```

```
./wrk  -t 1 -c 1 http://192.168.203.5:8000/testbucket/wrk
Running 10s test @ http://192.168.203.5:8000/testbucket/wrk
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   121.23us  189.31us   1.97ms   90.76%
    Req/Sec    14.21k     7.10k   27.77k    74.26%
  142804 requests in 10.10s, 85.66MB read
Requests/sec:  14139.52
Transfer/sec:      8.48MB
```

**Reply by avikivity** [reply]
Very encouraging!

**Comment by harshavardhana** [comment]
Latest release is 

```
./wrk  -t 1 -c 1 http://192.168.203.5:9001/testbucket/wrk
Running 10s test @ http://192.168.203.5:9001/testbucket/wrk
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   226.54us  169.89us   5.07ms   98.24%
    Req/Sec     4.69k   257.58     5.34k    72.00%
  46638 requests in 10.00s, 43.77MB read
Requests/sec:   4663.73
Transfer/sec:      4.38MB
```

More to come

**Reply by avikivity** [reply]
Impressive!
